middlewares/osio/jwt_token.go

```diff
@@ -0,0 +1,129 @@
+package middlewares
+
+import (
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/containous/traefik/log"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/pkg/errors"
+	"gopkg.in/square/go-jose.v1"
+)
+
+type PublicKey struct {
+	KeyID string
+	Key   *rsa.PublicKey
+}
+
+// JSONKeys the remote keys encoded in a json document
+type JSONKeys struct {
+	Keys []interface{} `json:"keys"`
+}
+
+func CreateSrvAccTokenChecker(client *http.Client, authURL string) SrvAccTokenChecker {
+	var publicKeysMap map[string]*rsa.PublicKey
+
+	keyFunc := func(token *jwt.Token) (interface{}, error) {
+		kid := token.Header["kid"]
+		if kid == nil {
+			log.Error("There is no 'kid' header in the token")
+			return nil, errors.New("There is no 'kid' header in the token")
+		}
+		key := publicKeysMap[fmt.Sprintf("%s", kid)]
+		if key == nil {
+			log.Error("There is no public key with such ID")
+			return nil, errors.New(fmt.Sprintf("There is no public key with such ID: %s", kid))
+		}
+		return key, nil
+	}
+
+	return func(token string) (bool, error) {
+		if publicKeysMap == nil {
+			remoteKeys, err := fetchKeys(client, authURL)
+			if err != nil {
+				return false, err
+			}
+			publicKeysMap = make(map[string]*rsa.PublicKey)
+			for _, remoteKey := range remoteKeys {
+				publicKeysMap[remoteKey.KeyID] = remoteKey.Key
+			}
+		}
```

Comment:
on first call to SrvAccTokenChecker(), it will fetch public_keys from auth service and then cache it forever.